lib/compute_visible_services.ts

```diff
@@ -0,0 +1,127 @@
+import * as documentDbUtils from "../lib/utils/documentdb";
+
+import { IContext } from "azure-function-express";
+import { DocumentClient } from "documentdb";
+import { isLeft } from "fp-ts/lib/Either";
+import { isNone, isSome } from "fp-ts/lib/Option";
+import { pick } from "italia-ts-commons/lib/types";
+import { RetrievedService } from "./models/service";
+import {
+  VISIBLE_SERVICE_COLLECTION_NAME,
+  VisibleServiceModel
+} from "./models/visible_service";
+import { getRequiredStringEnv } from "./utils/env";
+import { PermanentError } from "./utils/errors";
+import { configureAzureContextTransport } from "./utils/logging";
+
+// Whether we're in a production environment
+const isProduction = process.env.NODE_ENV === "production";
+
+const cosmosDbUri = getRequiredStringEnv("CUSTOMCONNSTR_COSMOSDB_URI");
+const cosmosDbKey = getRequiredStringEnv("CUSTOMCONNSTR_COSMOSDB_KEY");
+const cosmosDbName = getRequiredStringEnv("COSMOSDB_NAME");
+
+const documentDbDatabaseUrl = documentDbUtils.getDatabaseUri(cosmosDbName);
+
+const visibleServicesCollectionUrl = documentDbUtils.getCollectionUri(
+  documentDbDatabaseUrl,
+  VISIBLE_SERVICE_COLLECTION_NAME
+);
+const documentClient = new DocumentClient(cosmosDbUri, {
+  masterKey: cosmosDbKey
+});
+
+const visibleServiceModel = new VisibleServiceModel(
+  documentClient,
+  visibleServicesCollectionUrl
+);
+
+/**
+ * A function to cache the list of visible (active) services.
+ * Reacts to CosmosDB changes (change feed) in the services collection.
+ */
+export async function index(
+  context: IContext<{}>,
+  documents: ReadonlyArray<RetrievedService>
+): Promise<void> {
+  const logLevel = isProduction ? "info" : "debug";
+  configureAzureContextTransport(context, winston, logLevel);
+
+  try {
+    documents.forEach(async createdOrUpdatedService => {
+      winston.debug(
+        "ComputeVisibleService: %s",
+        JSON.stringify(createdOrUpdatedService)
+      );
+      const errorOrMaybeVisibleService = await visibleServiceModel.find(
+        createdOrUpdatedService.id,
+        createdOrUpdatedService.serviceId
+      );
+      if (isLeft(errorOrMaybeVisibleService)) {
+        winston.debug(
+          "ComputeVisibleService Query Error: %s",
+          JSON.stringify(errorOrMaybeVisibleService.value)
+        );
+        return PermanentError("QueryError querying for visible service");
+      }
+      const maybeVisibleService = errorOrMaybeVisibleService.value;
+      if (
+        !createdOrUpdatedService.isVisible &&
+        isSome(maybeVisibleService) &&
+        createdOrUpdatedService.version > maybeVisibleService.value.version
+      ) {
+        winston.debug(
+          "ComputeVisibleService: delete %s",
+          JSON.stringify(maybeVisibleService.value)
+        );
+        // if the service is not visible anymore
+        // delete it from the visible services list
+        await visibleServiceModel.delete(maybeVisibleService.value._self);
```

Comment:
> I thought we decided to keep the cache of visible services in a blob

yes but that's the approach that works well with batch scheduled calls (to compute the collection reacting to a timer trigger). if we use the cosmosdb change feed approach it is better to avoid a (very likely) race condition that may occur when updating the json blob (which is a problem that we don't have if we write to a collection). 

> the inefficient part if scanning the shards, not filtering the objects...

right but you can store visible services in an unpartioned collection of fixed size (up to 10GB) which solves the issue

to use a blob storage with this approach we should use some lock / release system (with retries) which would be somewhat complex to implement

Reply:
another option is to cache visible services into a redis set (which avoids to have locks on the shared json)

Reply:
> is better to avoid a (very likely) race condition that may occur when updating the json blob 

this is becoming complicated, why don't we just trigger the function every 10 minutes? in this case we're "almost" sure there will not be any race condition...

also we can simply reply on the "blob" output bindings of the function to update the blob, so it becomes very very simple

Reply:
ok